File: problems/1046/solution.py
from heapq import heapify, heappush, heappop

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        for i in range(len(stones)):
            stones[i] *= -1
        
        while len(stones) > 1:
            heapify(stones)
            
            s1 = heappop(stones)
            s2 = heappop(stones)
            
            if s1 != s2:
                stones.append(s1 - s2)
        
        return stones[0] * -1 if stones else 0

File: problems/1046/solution.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq;
        
        for (const int& stone: stones) {
            pq.push(stone);
        }
        
        while (size(pq) > 1) {
            int y = pq.top();
            pq.pop();
            int x = pq.top();
            pq.pop();
            
            if (x != y) {
                pq.push(y - x);
            }
        }
        
        return empty(pq) ? 0 : pq.top();
    }
};